onnxruntime/core/providers/rocm/roctracer_manager.cc

```diff
@@ -0,0 +1,465 @@
+  std::lock_guard<std::mutex> lock(roctracer_manager_mutex_);
+  auto res = next_client_id_++;
+  per_client_events_by_ext_correlation_.insert({res, {}});
+  ++num_active_clients_;
+  return res;
+}
+
+void RoctracerManager::DeregisterClient(uint64_t client_handle) {
+  std::lock_guard<std::mutex> lock(roctracer_manager_mutex_);
+  per_client_events_by_ext_correlation_.erase(client_handle);
+  --num_active_clients_;
+  if (num_active_clients_ == 0) {
+    StopLogging();
+  }
+}
+
+void RoctracerManager::StartLogging() {
+  std::lock_guard<std::mutex> lock(roctracer_manager_mutex_);
+  if (logging_enabled_) {
+    return;
+  }
+
+  // The following line shows up in all the samples, I do not know
+  // what the point is, but without it, the roctracer APIs don't work.
+  roctracer_set_properties(ACTIVITY_DOMAIN_HIP_API, nullptr);
+
+  roctracer_properties_t hcc_cb_properties;
+  memset(&hcc_cb_properties, 0, sizeof(roctracer_properties_t));
+  hcc_cb_properties.buffer_size = kActivityBufferSize;
+  hcc_cb_properties.buffer_callback_fun = ActivityCallback;
+  roctracer_open_pool(&hcc_cb_properties);
+
+  // Enable selective activity and API callbacks for the HIP APIs
+  roctracer_disable_domain_callback(ACTIVITY_DOMAIN_HIP_API);
+  roctracer_disable_domain_activity(ACTIVITY_DOMAIN_HIP_API);
+
+  for (auto const& logged_api : hip_api_calls_to_trace) {
+    uint32_t cid = 0;
+    roctracer_op_code(ACTIVITY_DOMAIN_HIP_API, logged_api.c_str(), &cid, nullptr);
+    roctracer_enable_op_callback(ACTIVITY_DOMAIN_HIP_API, cid, ApiCallback, nullptr);
+    roctracer_enable_op_activity(ACTIVITY_DOMAIN_HIP_API, cid);
+  }
+
+  // Enable activity logging in the HIP_OPS/HCC_OPS domain.
+  roctracer_enable_domain_activity(ACTIVITY_DOMAIN_HIP_OPS);
+
+  roctracer_start();
+  logging_enabled_ = true;
+}
+
+// Requires: roctracer_manager_mutex_ must be held
+void RoctracerManager::Clear() {
+  unprocessed_activity_buffers_.clear();
+  api_call_args_.clear();
+  unique_correlation_id_to_client_offset_.clear();
+  roctracer_correlation_to_unique_correlation_.clear();
+  per_client_events_by_ext_correlation_.clear();
+}
+
+// Requires: roctracer_manager_mutex_ must be held
+void RoctracerManager::StopLogging() {
+  if (!logging_enabled_) {
+    return;
+  }
+
+  roctracer_disable_domain_activity(ACTIVITY_DOMAIN_HIP_API);
+  roctracer_disable_domain_activity(ACTIVITY_DOMAIN_HIP_OPS);
+  roctracer_disable_domain_callback(ACTIVITY_DOMAIN_HIP_API);
+  roctracer_stop();
+  roctracer_flush_activity();
+  roctracer_close_pool();
+
+  logging_enabled_ = false;
+  Clear();
+}
+
+void RoctracerManager::Consume(uint64_t client_handle, const TimePoint& start_time,
+                               std::map<uint64_t, Events>& events) {
+  events.clear();
+  {
+    // Flush any pending activity records before starting
+    // to process the accumulated activity records.
+    std::lock_guard<std::mutex> lock_manager(roctracer_manager_mutex_);
+    roctracer_flush_activity();
```

Comment:
`RoctracerManager::Consume` is called in `RocmProfiler::EndProfiling`, and by [docstring](https://github.com/ROCm-Developer-Tools/roctracer/blob/ac64dff73d32a2f87d022a836b53d0f9598403e0/inc/roctracer.h#L743-L745)

>  If flushing encounters an activity record still being written, flushing stops. Use a subsequent flush when the record has completed being written to resume the flush.

IIUC, you need to flush until you see record_tp >= end_profiling_tp or maybe some other signal that indicate the pending activity records are flushed out.

Reply:
Yes, this is a valid point. However, the roctracer API is lacking, in that the timestamps used by roctracer do not have the same epoch as the userspace clocks. I'm presuming that it uses `ktime_get_ns` or some other kernel timestamp that measures time from system boot, rather than from a fixed epoch like Jan 1 1970. Getting these (kernel/ktime) timestamps from the userspace is a bit challenging and will probably involve additional dependencies on the kernel headers at least. Not sure that we want to do that.

Further, even if we did somehow get access to ktime timestamps, it's not necessary that we'll actually see a `record_timestamp >= end_profiling_timstamp` condition. The records are generated _in response_ to HIP API calls. So, if there are no HIP API calls after the call to `EndProfiling` (very plausible, considering that `EndProfiling` is naturally called at the end of a run), then we'll end up waiting forever.

So, in effect, yes, there's a small chance that the results might be incomplete in some cases, but I can't really think of a clean way to fix this problem.

Reply:
OK, it seems that this will be unsolvable open question. Lets just ensure a device sync is done before `EndProfiling`, in ORT python side. Also, it wont be a big problem is only some of the trailing events are missing, because we generally only care about the intermidiate batch step ;)